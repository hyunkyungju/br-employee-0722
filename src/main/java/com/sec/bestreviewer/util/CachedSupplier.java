package com.sec.bestreviewer.util;

import java.util.function.Supplier;

public class CachedSupplier<T> implements Supplier<T> {
    private final Supplier<T> wrapped;
    private volatile T cache;

    public CachedSupplier(Supplier<T> wrapped) {
        this.wrapped = wrapped;
    }

    @Override
    public T get() {
        if (cache == null) {
            cache = wrapped.get();
        }
        return cache;

    }
}