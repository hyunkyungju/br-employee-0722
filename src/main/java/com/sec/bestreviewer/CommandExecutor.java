package com.sec.bestreviewer;

import com.sec.bestreviewer.command.Command;
import com.sec.bestreviewer.store.EmployeeStore;
import com.sec.bestreviewer.store.Injection;
import com.sec.bestreviewer.util.CachedSupplier;

import java.util.List;

public class CommandExecutor {

    public static final int MAX_RESULT_NUMBER = 5;

    private final CachedSupplier<EmployeeStore> employeeStore;

    CommandExecutor() {
        employeeStore = new CachedSupplier<>(Injection::provideEmployeeStore);
    }

    CommandExecutor(final EmployeeStore employeeStore) {
        this.employeeStore = new CachedSupplier<>(() -> employeeStore);
    }

    List<String> execute(final Command command) {
        return command.execute(employeeStore.get());
    }
}