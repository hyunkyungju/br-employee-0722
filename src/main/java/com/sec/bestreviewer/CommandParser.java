package com.sec.bestreviewer;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class CommandParser {
    private final static int MIN_TOKENS_NUM = 5; // CMD + 3 of options + at least 1 params
    private final static String SPACE = " ";
    private final static String EMPTY = "";

    public TokenGroup parse(final String line) throws IllegalArgumentException {
        final String[] splits = line.split(",", -1);

        if (splits.length < MIN_TOKENS_NUM) {
            throw new IllegalArgumentException("wrong command format");
        }

        final String type = splits[0];
        final List<String> options = Arrays.asList(Arrays.copyOfRange(splits, 1, 4));
        final List<String> params = Arrays.asList(Arrays.copyOfRange(splits, 4, splits.length));

        return new TokenGroup(type, getValidList(options), getValidList(params));
    }

    private List<String> getValidList(final List<String> list) {
        final List<String> validList = new ArrayList<>();

        list.forEach(i -> {
            if (!EMPTY.equals(i) && !SPACE.equals(i)) {
                validList.add(i);
            }
        });

        return validList;
    }
}
