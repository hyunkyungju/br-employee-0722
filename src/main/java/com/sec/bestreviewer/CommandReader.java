package com.sec.bestreviewer;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class CommandReader {
    String inputFileName;

    CommandReader(final String inputFileName) {
        this.inputFileName = inputFileName;
    }

    public List<String> readFile() {
        final List<String> list = new ArrayList<>();
        try (final BufferedReader bufferedReader = new BufferedReader(new FileReader(inputFileName))) {
            for (String line = bufferedReader.readLine(); line != null; line = bufferedReader.readLine()) {
                list.add(line);
            }
        } catch (FileNotFoundException exception) {
            throw new IllegalArgumentException("Input file " + inputFileName + " is NOT exist");
        } catch (IOException exception) {
            exception.printStackTrace();
        }

        return list;
    }
}
