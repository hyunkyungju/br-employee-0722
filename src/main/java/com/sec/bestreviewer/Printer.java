package com.sec.bestreviewer;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Printer {
    private final String outputFilName;
    private final StringBuilder stdStringBuilder;

    public Printer(final String outputFilename) {
        this(outputFilename, new StringBuilder());
    }

    public Printer(final String outputFilename, final StringBuilder stdStringBuilder) {
        this.outputFilName = outputFilename;
        this.stdStringBuilder = stdStringBuilder;
    }

    public void add(final List<String> lines) {
        for (final String line : lines) {
            stdStringBuilder.append(line + "\n");
        }
    }

    public void printOutputFile() {
        try (OutputStream output = Files.newOutputStream(Paths.get(outputFilName))) {
            output.write(stdStringBuilder.toString().getBytes());
        } catch (IOException e) {
            e.getStackTrace();
        }
    }
}
