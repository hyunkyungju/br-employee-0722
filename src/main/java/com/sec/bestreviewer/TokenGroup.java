package com.sec.bestreviewer;

import lombok.AllArgsConstructor;
import lombok.Getter;

import java.util.List;

@Getter
@AllArgsConstructor
public class TokenGroup {
    private final String type;
    private final List<String> options;
    private final List<String> params;

}
