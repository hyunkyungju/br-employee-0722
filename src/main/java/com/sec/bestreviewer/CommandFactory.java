package com.sec.bestreviewer;

import com.sec.bestreviewer.command.*;
import com.sec.bestreviewer.condition.Condition;
import com.sec.bestreviewer.condition.ConditionFactory;
import com.sec.bestreviewer.model.EmployeeField;
import com.sec.bestreviewer.model.Employee;
import com.sec.bestreviewer.util.OptionParser;

import java.util.List;

public class CommandFactory {
    public static final String CMD_ADD = "ADD";
    public static final String CMD_DEL = "DEL";
    public static final String CMD_SCH = "SCH";
    public static final String CMD_CNT = "CNT";
    public static final String CMD_MOD = "MOD";

    static Command buildCommand(final String cmd, final List<String> options, final List<String> params) throws IllegalArgumentException {
        final OptionParser optionParser = new OptionParser(options);

        switch (cmd) {
            case CMD_ADD:
                final Employee employee =
                        Employee.of(params.get(EmployeeField.EMPLOYEE_NUMBER.ordinal()), params.get(EmployeeField.NAME.ordinal()),
                                params.get(EmployeeField.CAREER_LEVEL.ordinal()), params.get(EmployeeField.PHONE_NUMBER.ordinal()),
                                params.get(EmployeeField.BIRTHDAY.ordinal()), params.get(EmployeeField.CERTI.ordinal()));
                return new AddCommand(optionParser, employee);
            case CMD_DEL: {
                final Condition condition = ConditionFactory.createCondition(options, params, CMD_DEL);
                return new DeleteCommand(optionParser, condition);
            }
            case CMD_SCH: {
                final Condition condition = ConditionFactory.createCondition(options, params, CMD_SCH);
                return new SearchCommand(optionParser, condition);
            }
            case CMD_MOD: {
                final Condition condition = ConditionFactory.createCondition(options, params, CMD_MOD);
                final String changeColumn = params.get(params.size() - 2);
                final String changeValue = params.get(params.size() - 1);
                return new ModifyCommand(optionParser, condition, changeColumn, changeValue);
            }
            case CMD_CNT:
                return new CountCommand();
        }
        throw new IllegalArgumentException("Wrong command");
    }
}