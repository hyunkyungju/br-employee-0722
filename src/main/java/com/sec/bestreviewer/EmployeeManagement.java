package com.sec.bestreviewer;

import com.sec.bestreviewer.command.Command;
import com.sec.bestreviewer.condition.TokenGroup;

import java.util.Collections;
import java.util.List;

public class EmployeeManagement  {
    public void run(final String[] args) {
        if (args.length < 2) {
            printUsage();
            throw new IllegalArgumentException("wrong arguments count");
        }

        final String inputFileName = args[0];
        final String outputFilename = args[1];

        final CommandReader commandReader = new CommandReader(inputFileName);
        final CommandParser commandParser = new CommandParser();
        final CommandFactory commandFactory = new CommandFactory();
        final CommandExecutor commandExecutor = new CommandExecutor();
        final Printer printer = new Printer(outputFilename);

        final List<String> input = commandReader.readFile();

        for (final String line : input) {
            try {
                final TokenGroup tokens = commandParser.parse(line);
                final Command command = commandFactory.buildCommand(tokens.getType(), tokens.getOptions(), tokens.getParams(), tokens.getCondTokens());
                final List<String> result = commandExecutor.execute(command);
                printer.add(result);
            } catch (IllegalArgumentException exception) {
                printer.add(Collections.singletonList("wrong command : " + line));
            }
        }

        printer.printOutputFile();
    }

    private void printUsage() {
        System.out.println("input / output 형태는 txt 파일이며, 아래 command 로 input file 을 read 하여, output file 을 생성한다.");
        System.out.println("Usage : EmployeeManagement input.txt output.txt");
    }

    public static void main(final String[] args) {
        final EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);
    }
}