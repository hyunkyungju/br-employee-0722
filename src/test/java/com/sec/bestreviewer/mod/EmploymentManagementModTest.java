package com.sec.bestreviewer.mod;

import com.sec.bestreviewer.EmployeeManagement;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;

public class EmploymentManagementModTest {

    private Path inputFile;
    private Path outputFile;
    private String[] args;

    @BeforeEach
    void setUp() throws IOException {

        String input = String.join("\n", List.of(
                "ADD, , , ,91351446,LIM PNQN,CL3,010-1234-5678,19700122,ADV",
                "ADD, , , ,06134213,JIN LIN,CL3,010-0320-1235,19821022,PRO",
                "ADD, , , ,09143542,HAEIN JUNG,CL3,010-0210-5423,19840212,EX",
                "ADD, , , ,10142536,MICHAEL JACKSON,CL3,010-0410-4532,19740730,ADV",
                "ADD, , , ,11231432,AMANDI GUPTA,CL3,010-0650-1362,19880228,EX",
                "ADD, , , ,18350301,KIM EOE,CL4,010-0840-4988,19810408,EX",
                "ADD, , , ,07843022,SEO KFI,CL3,010-0830-6716,19810408,PRO",
                "ADD, , , ,12345678,YUJIN KIM,CL2,010-0970-0055,19900101,PRO",
                "ADD, , , ,23456789,YUJIN LEE,CL2,010-0970-0055,19920101,PRO",
                "ADD, , , ,34567890,MIN KIM,CL2,010-0970-0066,19900106,PRO",
                "ADD, , , ,45678901,LEE YUJIN,CL1,010-1111-2222,19901225,EX",
                "ADD, , , ,56789012,PARK KIM,CL1,010-1111-2223,19901206,PRO",
                "ADD, , , ,67890123,YUJIN KIM,CL1,010-1111-2224,19900106,ADV",
                "ADD, , , ,78901234,KIM YUJIN,CL2,010-9999-8888,19880909,EX",
                "ADD, , , ,89012345,YUJIN PARK,CL3,010-9999-8888,19871007,PRO",
                "ADD, , , ,90123456,UNKNOWN NAME,CL1,010-0000-0000,19990101,ADV")) + "\n";

        inputFile = Files.createTempFile("test-data", ".txt");
        outputFile = Files.createTempFile("test-output", ".txt");
        args = new String[]{
                inputFile.toFile().getAbsolutePath(),
                outputFile.toFile().getAbsolutePath()
        };

        Files.writeString(inputFile, input);
    }

    @AfterEach
    void tearDown() throws IOException {
        Files.deleteIfExists(inputFile);
        Files.deleteIfExists(outputFile);
    }

    @DisplayName("[MOD] 사용자 이름 변경")
    @Test
    void modChangeName() throws IOException {
        String command = "MOD,-p, , ,name,YUJIN KIM,name,YUJIN LEE\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] CL로 검색 & 핸드폰 번호 변경")
    @Test
    void modChangePhoneNumberByCL() throws IOException {
        String command = "MOD, , , ,cl,CL3,phoneNum,010-0970-0055\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 핸드폰 번호 검색 & 생년월일 변경")
    @Test
    void modChangeBirthdayByPhoneNumber() throws IOException {
        String command = "MOD,-p, , ,phoneNum,010-0970-0055,birthday,20001225\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 성(Last Name)으로 전화번호 변경")
    @Test
    void modChangePhoneNumberByLastName() throws IOException {
        String command = "MOD,-p,-l, ,name,KIM,phoneNum,010-0970-0055\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 전화번호 뒷자리로 certi 변경")
    @Test
    void modChangeCertiByPhoneNumberTail() throws IOException {
        String command = "MOD, ,-l, ,phoneNum,0055,certi,EX\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 전화번호 중간으로 certi 변경")
    @Test
    void modChangeCertiByPhoneNumberMiddle() throws IOException {
        String command = "MOD, ,-m, ,phoneNum,0970,certi,EX\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 생년월일 연도로 이름 변경")
    @Test
    void modChangeNameByBirthYear() throws IOException {
        String command = "MOD,-p,-y, ,birthday,1990,name,YUJIN LEE\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] and 옵션: CL2 + 생년월 01로 이름 변경")
    @Test
    void modChangeNameByCLAndBirthMonth() throws IOException {
        String command = "MOD,-p, , ,cl,CL2,-a,-m, ,birthday,01,name,YUJIN LEE\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] or 옵션: 생년일 06 또는 certi PRO로 생년월일 변경")
    @Test
    void modChangeBirthdayByBirthDayOrCerti() throws IOException {
        String command = "MOD, ,-d, ,birthday,06,-o, , ,certi,PRO,birthday,19901225\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 조건 불일치(NONE)")
    @Test
    void modNone() throws IOException {
        String command = "MOD,-p, , ,employeeNum,99999999,name,NEW NAME\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }

    @DisplayName("[MOD] 사원번호 수정 불가")
    @Test
    void modChangeEmployeeNumNotAllowed() throws IOException {
        String command = "MOD, , , ,employeeNum,91351446,employeeNum,00000000\n";
        Files.writeString(inputFile, command, StandardOpenOption.APPEND);

        EmployeeManagement employeeManagement = new EmployeeManagement();
        employeeManagement.run(args);

        String output = Files.readString(outputFile);
        Approvals.verify(output);
    }
}
