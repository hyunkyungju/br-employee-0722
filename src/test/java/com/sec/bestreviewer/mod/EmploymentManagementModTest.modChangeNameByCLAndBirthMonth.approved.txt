MOD,12345678,YUJIN KIM,CL2,010-0970-0055,19900101,PRO
MOD,23456789,YUJIN LEE,CL2,010-0970-0055,19920101,PRO
MOD,34567890,MIN KIM,CL2,010-0970-0066,19900106,PRO